mpi: autogen errhan api

hzhou opened the pull request:
## Pull Request Description

[skip warnings]

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
Rebased:

test:mpich/ch3/most
test:mpich/ch4/most

shintaro-iwasaki reviewed (commented):
Thanks! Overall it looks good. Except for minor error checks, the main changes I would appreciate are:
1. Create another commit that implements `MPIR_Errhandler_free_impl()` before this change.
2. Create another commit that moves this `*errorclass > MPIR_Process.attrs.lastusedcode` check.

If we want to remove `ThreadSafeNoUpdate`, I want another commit that explains why we remove `ThreadSafeNoUpdate`.

hzhou commented on src/mpi/errhan/errhan_impl.c:
```diff
@@ -123,60 +125,51 @@ void MPIR_File_get_errhandler_impl(MPI_File file, MPI_Errhandler * errhandler)
-        if (!in_use) {
-            MPIR_Errhandler_free(comm_ptr->errhandler);
-        }
+        MPIR_Errhandler_free_impl(comm_ptr->errhandler);
```
Separated commit https://github.com/pmodels/mpich/pull/4957/commits/acc703dce4628e6245e9764974c59235b35a8cb4 , which includes refactoring of `errhan_impl.c`.

hzhou commented on src/mpi/errhan/dynerrutil.c:
```diff
@@ -289,7 +288,12 @@ int MPIR_Err_add_class(void)
 
-    return (new_class | ERROR_DYN_MASK);
+    new_class |= ERROR_DYN_MASK;
+    if (new_class > MPIR_Process.attrs.lastusedcode) {
```
Separated commit https://github.com/pmodels/mpich/pull/4957/commits/38aee9aca1c8fa7b29395d7a6a8b59fa9f0087a4 for all error utility impl functions.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .desc: Call the error handler installed on a file
+
+MPI_File_create_errhandler:
+    .desc: Create a file error handler
```
Fixed.

hzhou commented on src/mpi/errhan/errhan_impl.c:
```diff
@@ -64,18 +63,21 @@ int MPIR_File_create_errhandler_impl(MPI_File_errhandler_function * file_errhand
-}
+#else
+    *errhandler_ptr = NULL;
+    return MPI_SUCCESS;
```
Fixed.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .extra: ignore_revoked_comm
+
+MPI_File_call_errhandler:
+    .desc: Call the error handler installed on a file
```
Added `.docnotes: ThreadSafeNoUpdate` for all three `_call_errhandler` functions. Updated the python scripts to skip global critical section.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+
+MPIX_Delete_error_class:
+    errorclass: ERROR_CLASS, [value of the error class to be remoed]
+    .desc: Delete an MPI error class from the known classes
```
`MPIR_ERRTEST_ARGNULL` checks whether a *pointer* argument is `NULL`. It shouldn't be applied to `errorclass`. The implementation (`MPIR_Delete_error_class_impl`) will throw error when the `errorclass` is wrong (missing in hash) so it is OK to skip the validation here.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+
+MPIX_Delete_error_code:
+    errorcode: ERROR_CODE, [value of the error code to be remoed]
+    .desc: Delete an MPI error code
```
Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548619551

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    for an error code or class with a value less than or equal
+    to 'MPI_ERR_LASTCODE'.  Thus, you cannot replace the predefined error messages
+    with this routine.
+*/
```
Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548619551

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .seealso: MPI_Comm_create_errhandler, MPI_Errhandler_free
+
+MPI_Errhandler_free:
+    .desc: Frees an MPI-style errorhandler
```
`errhandler` is an "inout" argument. I guess we should check for NULL before we access it with `*errhandler`. This applies to all `inout` handle argument. Let me add the rule to python script.

... Added.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+
+MPI_Comm_call_errhandler:
+    .desc: Call the error handler installed on a communicator
+    .extra: ignore_revoked_comm
```
> 
> 
> Previously, this function is "ThreadSafeNoUpdate" and does not take a lock.
> Now this routine takes a lock (`MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX`) and becomes "ThreadSafe".
> Is this change intended and acceptable?
> 
> Even if there's a good reason to take a lock, I want to see a separate commit that explains this change.

Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548602715

hzhou replied:
> 
> 
> Don't we need to check the following?
> 
> ```c
>             if (comm_ptr->errhandler) {
>                 MPIR_ERRTEST_ERRHANDLER(comm_ptr->errhandler->handle, mpi_errno);
>             }
> ```

The handle that is attached to comm (or win, or file) is always valid, or it won't get attached in the first place. So no need to check

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .extra: ignore_revoked_comm
+
+MPI_Comm_create_errhandler:
+    .desc: Create a communicator error handler
```
Added rule to the python script

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+MPI_Comm_set_errhandler:
+    .desc: Set the error handler for a communicator
+    .seealso: MPI_Comm_get_errhandler, MPI_Comm_call_errhandler
+    .extra: ignore_revoked_comm
```
Builtin errhandler will also pass `MPIR_Errhandler_valid_ptr`.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .desc: Get the error handler attached to a file
+
+MPI_File_set_errhandler:
+    .desc: Set the error handler for an MPI file
```
> ThreadSafeNoUpdate

Added `.docnotes: ThreadSafeNoUpdate`

hzhou replied:
> ```c
> MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__,
>                                                  __LINE__, MPI_ERR_ARG, "**errhandnotfile", NULL);
> ```

Added custom error_check in `errhan_api.txt`.

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .desc: Set the error handler for an MPI file
+
+MPI_Win_call_errhandler:
+    .desc: Call the error handler installed on a window
```
Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548602715

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .desc: Call the error handler installed on a window
+
+MPI_Win_create_errhandler:
+    .desc: Create an error handler for use with MPI window
```
Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548635241

hzhou commented on src/binding/c/errhan_api.txt:
```diff
@@ -0,0 +1,157 @@
+    .desc: Get the error handler for the MPI RMA window
+
+MPI_Win_set_errhandler:
+    .desc: Set window error handler
```
Added custom error_check in `errhan_api.txt`

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

shintaro-iwasaki reviewed (approved):
@hzhou.  Thanks a lot!  I approve this PR assuming that 1. you will fix the following and 2. pass the tests before merging this PR unless there's a special reason to leave them as `ThreadSafe`.

- Add `.ThreadSafeNoUpdate`
  - `MPI_Comm_get_errhandler()`
  - `MPI_Comm_set_errhandler()`
  - `MPI_File_get_errhandler()`
  - `MPI_Win_set_errhandler()`
  - (`MPI_Win_get_errhandler()`)

`MPI_Win_get_errhandler()` is not marked as `ThreadSafeNoUpdate`, but because `MPI_Win_set_errhandler()` and `MPI_Win_call_errhandler()` are `ThreadSafeNoUpdate`, this getter should also be `ThreadSafeNoUpdate`.

Sorry for missing them in the previous review.

hzhou commented:
> 
> 
> @hzhou. Thanks a lot! I approve this PR assuming that 1. you will fix the following and 2. pass the tests before merging this PR unless there's a special reason to leave them as `ThreadSafe`.
> 
>     * Add `.ThreadSafeNoUpdate`
>       
>       * `MPI_Comm_get_errhandler()`
>       * `MPI_Comm_set_errhandler()`
>       * `MPI_File_get_errhandler()`
>       * `MPI_Win_set_errhandler()`
>       * (`MPI_Win_get_errhandler()`)
> 
> 
> `MPI_Win_get_errhandler()` is not marked as `ThreadSafeNoUpdate`, but because `MPI_Win_set_errhandler()` and `MPI_Win_call_errhandler()` are `ThreadSafeNoUpdate`, this getter should also be `ThreadSafeNoUpdate`.
> 
> Sorry for missing them in the previous review.

I added `.ThreadSafeNoUpdate` note to these functions just to be consistent with current man pages. However, on seconding reading `maint/doctext`, I am not sure the intended meaning of it. I For the mentioned `info` case, since user is allowed to set info outside MPI `Init/Finailze` window, we can't really protect the info with global lock. I guess that is where this note is originally intended for. I am not sure what is the object supposed to be in the errorhandler case. It seems to me that all there functions can and should be protected by the lock. The bottom line is, I don't think `ThreadSafeNoUpdate` implies not taking global lock. Therefore, I removed the commit that skips `global_cs` automatically based on `ThreadSafeNoUPdate`. If we need skip `global_cs`, just do that separately.

Let me know what do you think.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

shintaro-iwasaki commented:
@hzhou I believe this PR should focus on translating the current mechanism as it is (except for very obvious issues) and let #4915 deal with a lock problem. In this sense, I would like to keep a global lock removed from `.ThreadSafeNoUpdate` functions.

hzhou commented:
> 
> 
> @hzhou I believe this PR should focus on translating the current mechanism as it is

The key issue is "the current mechanism" is inconsistent:
```
$ grep ThreadSafeNoUpdate *_errhandler.c
comm_call_errhandler.c:.N ThreadSafeNoUpdate
comm_get_errhandler.c:.N ThreadSafeNoUpdate
comm_set_errhandler.c:.N ThreadSafeNoUpdate
file_call_errhandler.c:.N ThreadSafeNoUpdate
file_get_errhandler.c:.N ThreadSafeNoUpdate
file_set_errhandler.c:.N ThreadSafeNoUpdate
win_call_errhandler.c:.N ThreadSafeNoUpdate
win_set_errhandler.c:.N ThreadSafeNoUpdate

$ grep _ENTER.*MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX *_errhandler.c
comm_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
comm_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
comm_set_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
file_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
file_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
win_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
win_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```

>  In this sense, I would like to keep a global lock removed from `.ThreadSafeNoUpdate` functions.
> 

By your argument, it is wrong "to keep a global lock removed from `.ThreadSafeNoUpdate` functions", right?

shintaro-iwasaki commented:
> "to keep a global lock removed from `.ThreadSafeNoUpdate` functions"

You're right. Sorry I misunderstood something. `.ThreadSafeNoUpdate` does not imply `.global_cs`. We need explicit `.global_cs` if necessary.

I am still afraid of significantly changing the behavior of functions in this PR (e.g., by adding a lock) since this PR is for auto-generation. Of course it is unclear whether the change is "changing the behavior" or "fixing inconsistency," but at least, `MPI_xxx_call_errhandler()` currently has no lock. If we can keep it removed for these functions (by adding `.global_cs`), I'm okay. If we will remove them, please create a different commit that explains this change.

hzhou commented:
Added `.skip: global_cs` to `MPI_{Comm,Win,File}_call_errhandler`.

test:mpich/ch3/most
test:mpich/ch4/most

shintaro-iwasaki commented:
@hzhou Thanks!
